File: R/01_data_audit_and_fix.R
# ============================================================================
# NMA Methods Research: Data Audit & Fix Script (Step 1)
# ============================================================================
# Purpose: Audit all datasets, identify issues, and create fix recipes
# ============================================================================

# Set working directory -- relative to this script's location
setwd(file.path(dirname(sys.frame(1)$ofile %||% "."), ".."))

# Load required packages
suppressPackageStartupMessages({
  library(dplyr)
  library(tidyr)
  library(readr)
})

cat("\n")
cat("=================================================================\n")
cat("  NMA DATA AUDIT & FIX - Step 1\n")
cat("=================================================================\n")
cat("\n")

# ============================================================================
# SECTION 1: Audit all CSV files
# ============================================================================

cat("Section 1: Auditing CSV files...\n\n")

# Define directories
nmadatasets_dir <- "nmadatasets/inst/extdata"
netmeta_dir <- "netmetaDatasets/inst/extdata"

# Get all CSV files
nmadatasets_files <- list.files(nmadatasets_dir, pattern = "\\.csv$", full.names = TRUE)
netmeta_files <- list.files(netmeta_dir, pattern = "\\.csv$", full.names = TRUE)

all_files <- c(nmadatasets_files, netmeta_files)

cat(sprintf("Found %d CSV files to audit\n\n", length(all_files)))

# Audit function
audit_csv <- function(filepath) {
  result <- list(
    file = basename(filepath),
    path = filepath,
    status = "OK",
    n_rows = 0,
    n_cols = 0,
    columns = character(),
    issues = character(),
    sample_data = NULL
  )

  tryCatch({
    df <- read.csv(filepath, stringsAsFactors = FALSE, nmax = 100)

    result$n_rows <- nrow(df)
    result$n_cols <- ncol(df)
    result$columns <- names(df)

    # Check for issues
    # 1. Check for NA values
    for (col in names(df)) {
      na_count <- sum(is.na(df[[col]]))
      if (na_count > 0) {
        result$issues <- c(result$issues,
          sprintf("Column '%s': %d NA values (%.1f%%)",
            col, na_count, 100*na_count/nrow(df)))
      }
    }

    # 2. Check for empty treatment names
    if ("treatment" %in% names(df)) {
      empty_treatments <- df$treatment == "" | df$treatment == " " | is.na(df$treatment)
      if (any(empty_treatments)) {
        result$issues <- c(result$issues,
          sprintf("%d empty treatment labels", sum(empty_treatments)))
      }

      # Check for placeholder labels
      placeholder_pattern <- "^[0-9]+$|^[A-D]$"
      if (any(grepl(placeholder_pattern, df$treatment))) {
        result$issues <- c(result$issues,
          "Contains placeholder treatment labels (numbers or A-D)")
      }
    }

    # 3. Check for empty study names
    if ("study" %in% names(df)) {
      empty_studies <- df$study == "" | df$study == " " | is.na(df$study)
      if (any(empty_studies)) {
        result$issues <- c(result$issues,
          sprintf("%d empty study labels", sum(empty_studies)))
      }

      # Check for placeholder study IDs
      placeholder_pattern <- "^[0-9]+$"
      if (any(grepl(placeholder_pattern, df$study))) {
        result$issues <- c(result$issues,
          "Contains numeric/placeholder study IDs")
      }
    }

    # 4. Check for duplicates
    if (all(c("study", "treatment") %in% names(df))) {
      n_dups <- sum(duplicated(df[, c("study", "treatment")]))
      if (n_dups > 0) {
        result$issues <- c(result$issues,
          sprintf("%d duplicate (study, treatment) rows", n_dups))
      }
    }

    # 5. Check if file is empty
    if (nrow(df) == 0) {
      result$status <- "EMPTY"
      result$issues <- c(result$issues, "File is empty")
    } else if (length(result$issues) > 0) {
      result$status <- "ISSUES"
    }

    # Store sample data
    result$sample_data <- head(df, 3)

  }, error = function(e) {
    result$status <- "ERROR"
    result$issues <- as.character(e$message)
  })

  return(result)
}

# Run audit on all files
audit_results <- lapply(all_files, audit_csv)

# Summarize by status
status_summary <- table(sapply(audit_results, function(x) x$status))

cat("Audit Summary:\n")
print(status_summary)
cat("\n")

# Group by package
cat("Files by Package:\n")
by_package <- sapply(audit_results, function(x) {
  if (grepl("nmadatasets", x$path)) {
    if (grepl("bnma__", x$file)) "bnma"
    else if (grepl("BUGSnet__", x$file)) "BUGSnet"
    else if (grepl("gemtc__", x$file)) "gemtc"
    else if (grepl("MBNMAdose__", x$file)) "MBNMAdose"
    else if (grepl("multinma__", x$file)) "multinma"
    else if (grepl("nmaINLA__", x$file)) "nmaINLA"
    else if (grepl("pcnetmeta__", x$file)) "pcnetmeta"
    else if (grepl("netmeta__", x$file)) "netmeta"
    else "nmadatasets_other"
  } else {
    "netmetaDatasets"
  }
})

package_summary <- table(by_package)
print(package_summary)
cat("\n")

# ============================================================================
# SECTION 2: Identify Cross-Package Datasets
# ============================================================================

cat("\nSection 2: Identifying Cross-Package Datasets...\n\n")

# Map file to dataset name
get_dataset_name <- function(filename) {
  # Extract dataset ID from filename
  # e.g., "bnma__smoking_studies.csv" -> "smoking"
  # e.g., "gemtc__thrombolytic_nodes.csv" -> "thrombolytic"

  if (grepl("__smoking", filename)) return("smoking")
  if (grepl("__thrombolyt", filename)) return("thrombolytic")
  if (grepl("__parkinson", filename)) return("parkinson")
  if (grepl("__diabetes", filename)) return("diabetes")
  if (grepl("__depression", filename)) return("depression")
  if (grepl("__dietfat|__dietary", filename)) return("dietary_fat")
  if (grepl("__blocker", filename)) return("blocker")
  if (grepl("__statins", filename)) return("statins")

  return(gsub("__.*$", "", filename))
}

# Group by dataset name
dataset_names <- sapply(audit_results, function(x) get_dataset_name(x$file))
dataset_presence <- table(dataset_names)

cat("Datasets appearing multiple times (cross-package):\n")
multi_presence <- dataset_presence[dataset_presence > 2]
print(multi_presence)
cat("\n")

# ============================================================================
# SECTION 3: Detailed Issue Report
# ============================================================================

cat("\nSection 3: Detailed Issue Report\n")
cat("==================================\n\n")

# Get files with issues
problematic <- Filter(function(x) x$status %in% c("ISSUES", "ERROR", "EMPTY"), audit_results)

cat(sprintf("Found %d files with issues\n\n", length(problematic)))

# Group by issue type
for (i in seq_along(problematic)) {
  item <- problematic[[i]]
  cat(sprintf("[%d] %s\n", i, item$file))
  cat(sprintf("    Status: %s\n", item$status))
  cat(sprintf("    Size: %d rows x %d cols\n", item$n_rows, item$n_cols))
  cat("    Issues:\n")
  for (issue in item$issues) {
    cat(sprintf("      - %s\n", issue))
  }
  cat("\n")
}

# ============================================================================
# SECTION 4: Generate Fix Recipes
# ============================================================================

cat("\nSection 4: Generating Fix Recipes\n")
cat("================================\n\n")

fix_recipes <- list()

for (item in problematic) {

  recipe <- list(
    file = item$file,
    path = item$path,
    fixes = character()
  )

  for (issue in item$issues) {
    if (grepl("placeholder treatment", issue)) {
      recipe$fixes <- c(recipe$fixes,
        "FIX: Load from source package and extract treatment labels")
    } else if (grepl("numeric/placeholder study IDs", issue)) {
      recipe$fixes <- c(recipe$fixes,
        "FIX: Load from source package and extract study names")
    } else if (grepl("duplicate", issue)) {
      recipe$fixes <- c(recipe$fixes,
        "FIX: Remove duplicate (study, treatment) rows, keep first")
    } else if (grepl("NA values", issue)) {
      recipe$fixes <- c(recipe$fixes,
        "FIX: Impute or remove rows with NA values")
    } else if (grepl("File is empty", issue)) {
      recipe$fixes <- c(recipe$fixes,
        "FIX: Re-extract from source package")
    }
  }

  if (length(recipe$fixes) > 0) {
    fix_recipes[[length(fix_recipes) + 1]] <- recipe
  }
}

cat(sprintf("Generated %d fix recipes\n\n", length(fix_recipes)))

# ============================================================================
# SECTION 5: Save Results
# ============================================================================

cat("Section 5: Saving Results\n")
cat("=========================\n\n")

# Create results directory
results_dir <- "outputs"
if (!dir.exists(results_dir)) {
  dir.create(results_dir, recursive = TRUE)
}

# Save audit summary
audit_summary <- data.frame(
  file = sapply(audit_results, function(x) x$file),
  status = sapply(audit_results, function(x) x$status),
  n_rows = sapply(audit_results, function(x) x$n_rows),
  n_cols = sapply(audit_results, function(x) x$n_cols),
  n_issues = sapply(audit_results, function(x) length(x$issues)),
  issues = I(sapply(audit_results, function(x) paste(x$issues, collapse = "; ")))
)

write.csv(audit_summary, file.path(results_dir, "data_audit_summary.csv"), row.names = FALSE)
cat("Saved: data_audit_summary.csv\n")

# Save problematic files detail
if (length(problematic) > 0) {
  problematic_detail <- data.frame(
    file = sapply(problematic, function(x) x$file),
    status = sapply(problematic, function(x) x$status),
    issues = I(sapply(problematic, function(x) paste(x$issues, collapse = "; ")))
  )
  write.csv(problematic_detail, file.path(results_dir, "problematic_files_detail.csv"), row.names = FALSE)
  cat("Saved: problematic_files_detail.csv\n")
}

# Save fix recipes
if (length(fix_recipes) > 0) {
  fix_recipes_df <- data.frame(
    file = sapply(fix_recipes, function(x) x$file),
    fixes = I(sapply(fix_recipes, function(x) paste(x$fixes, collapse = "; ")))
  )
  write.csv(fix_recipes_df, file.path(results_dir, "fix_recipes.csv"), row.names = FALSE)
  cat("Saved: fix_recipes.csv\n")
}

# Save cross-package dataset mapping
cross_package_mapping <- data.frame(
  dataset = names(dataset_presence),
  n_files = as.integer(dataset_presence)
)
cross_package_mapping <- cross_package_mapping[order(cross_package_mapping$n_files, decreasing = TRUE), ]
write.csv(cross_package_mapping, file.path(results_dir, "cross_package_datasets.csv"), row.names = FALSE)
cat("Saved: cross_package_datasets.csv\n")

cat("\n")
cat("=================================================================\n")
cat("  AUDIT COMPLETE\n")
cat("=================================================================\n")
cat(sprintf("\nResults saved to: %s\n", results_dir))
cat("\n")

# Print summary
cat("SUMMARY:\n")
cat(sprintf("  Total files audited: %d\n", length(audit_results)))
cat(sprintf("  Files with issues: %d\n", length(problematic)))
cat(sprintf("  Cross-package datasets: %d\n", sum(dataset_presence > 2)))
cat("\nNext steps:\n")
cat("  1. Review problematic_files_detail.csv\n")
cat("  2. Apply fixes from fix_recipes.csv\n")
cat("  3. Run cross-package comparison on clean datasets\n")
cat("\n")

File: R/01_data_audit_and_fix.py
#!/usr/bin/env python3
"""
NMA Methods Research: Data Audit & Fix Script (Step 1)
Purpose: Audit all datasets, identify issues, and create fix recipes
"""

import os
import csv
from collections import defaultdict
from pathlib import Path

# Set working directory -- relative to project root
BASE_DIR = Path(__file__).resolve().parent.parent
OUTPUT_DIR = BASE_DIR / "outputs"
NMADATASETS_DIR = BASE_DIR / "nmadatasets" / "inst" / "extdata"
NETMETA_DIR = BASE_DIR / "netmetaDatasets" / "inst" / "extdata"

print("\n" + "="*73)
print("  NMA DATA AUDIT & FIX - Step 1")
print("="*73 + "\n")

# ============================================================================
# SECTION 1: Audit all CSV files
# ============================================================================

print("Section 1: Auditing CSV files...\n")

# Get all CSV files
all_files = []
if NMADATASETS_DIR.exists():
    all_files.extend(list(NMADATASETS_DIR.glob("*.csv")))
if NETMETA_DIR.exists():
    all_files.extend(list(NETMETA_DIR.glob("*.csv")))

print(f"Found {len(all_files)} CSV files to audit\n")

# Audit function
def audit_csv(filepath):
    result = {
        'file': filepath.name,
        'path': str(filepath),
        'status': 'OK',
        'n_rows': 0,
        'n_cols': 0,
        'columns': [],
        'issues': [],
        'sample_data': []
    }

    try:
        with open(filepath, 'r', encoding='utf-8', errors='ignore') as f:
            reader = csv.DictReader(f)
            rows = list(reader)

            if not rows:
                result['status'] = 'EMPTY'
                result['issues'].append('File is empty')
                return result

            result['n_rows'] = len(rows)
            result['n_cols'] = len(rows[0]) if rows else 0
            result['columns'] = list(rows[0].keys()) if rows else []

            # Check for issues
            # 1. Check for empty values
            for col in result['columns']:
                na_count = sum(1 for row in rows if not row.get(col) or row.get(col).strip() == '')
                if na_count > 0:
                    pct = 100 * na_count / len(rows)
                    result['issues'].append(
                        f"Column '{col}': {na_count} empty values ({pct:.1f}%)"
                    )

            # 2. Check for empty treatment names
            if 'treatment' in result['columns']:
                empty_treatments = sum(1 for row in rows
                    if not row.get('treatment') or row.get('treatment').strip() in ('', 'NA'))
                if empty_treatments > 0:
                    result['issues'].append(f"{empty_treatments} empty treatment labels")

                # Check for placeholder labels
                import re
                placeholder_count = sum(1 for row in rows
                    if row.get('treatment') and re.match(r'^\d+$|^[A-D]$', row.get('treatment', '').strip()))
                if placeholder_count > 0:
                    result['issues'].append("Contains placeholder treatment labels (numbers or A-D)")

            # 3. Check for empty study names
            if 'study' in result['columns']:
                empty_studies = sum(1 for row in rows
                    if not row.get('study') or row.get('study').strip() in ('', 'NA'))
                if empty_studies > 0:
                    result['issues'].append(f"{empty_studies} empty study labels")

                # Check for placeholder study IDs
                numeric_study_count = sum(1 for row in rows
                    if row.get('study') and re.match(r'^\d+$', row.get('study', '').strip()))
                if numeric_study_count > 0:
                    result['issues'].append("Contains numeric/placeholder study IDs")

            # 4. Check for duplicates
            if 'study' in result['columns'] and 'treatment' in result['columns']:
                seen = set()
                dup_count = 0
                for row in rows:
                    key = (row.get('study', ''), row.get('treatment', ''))
                    if key in seen:
                        dup_count += 1
                    seen.add(key)
                if dup_count > 0:
                    result['issues'].append(f"{dup_count} duplicate (study, treatment) rows")

            # Store sample data
            result['sample_data'] = rows[:3]

            # Set status
            if len(result['issues']) > 0:
                result['status'] = 'ISSUES'

    except Exception as e:
        result['status'] = 'ERROR'
        result['issues'].append(str(e))

    return result


# Run audit
audit_results = [audit_csv(f) for f in all_files]

# Summarize by status
status_summary = defaultdict(int)
for r in audit_results:
    status_summary[r['status']] += 1

print("Audit Summary:")
for status, count in sorted(status_summary.items()):
    print(f"  {status}: {count}")
print()

# Group by package
def get_package(filename):
    if 'nmadatasets' in filename:
        if 'bnma__' in filename: return 'bnma'
        if 'BUGSnet__' in filename: return 'BUGSnet'
        if 'gemtc__' in filename: return 'gemtc'
        if 'MBNMAdose__' in filename: return 'MBNMAdose'
        if 'multinma__' in filename: return 'multinma'
        if 'nmaINLA__' in filename: return 'nmaINLA'
        if 'pcnetmeta__' in filename: return 'pcnetmeta'
        if 'netmeta__' in filename: return 'netmeta'
        return 'nmadatasets_other'
    return 'netmetaDatasets'

package_summary = defaultdict(int)
for r in audit_results:
    pkg = get_package(r['path'])
    package_summary[pkg] += 1

print("Files by Package:")
for pkg, count in sorted(package_summary.items(), key=lambda x: -x[1]):
    print(f"  {pkg}: {count}")
print()

# ============================================================================
# SECTION 2: Identify Cross-Package Datasets
# ============================================================================

print("\nSection 2: Identifying Cross-Package Datasets...\n")

def get_dataset_name(filename):
    'Extract dataset name from filename'
    filename_lower = filename.lower()
    if '__smoking' in filename_lower: return 'smoking'
    if '__thrombolyt' in filename_lower: return 'thrombolytic'
    if '__parkinson' in filename_lower: return 'parkinson'
    if '__diabetes' in filename_lower: return 'diabetes'
    if '__depression' in filename_lower: return 'depression'
    if '__dietfat' in filename_lower or '__dietary' in filename_lower: return 'dietary_fat'
    if '__blocker' in filename_lower: return 'blocker'
    if '__statins' in filename_lower: return 'statins'
    # Extract from package__dataset pattern
    if '__' in filename:
        parts = filename.split('__')
        if len(parts) > 1:
            return parts[1].replace('_nodes', '').replace('_studies', '').replace('_contrast', '')
    return filename.split('_')[0]

dataset_presence = defaultdict(int)
for r in audit_results:
    ds_name = get_dataset_name(r['file'])
    dataset_presence[ds_name] += 1

print("Datasets appearing multiple times (cross-package):")
multi_presence = {k: v for k, v in sorted(dataset_presence.items(), key=lambda x: -x[1]) if v > 2}
for ds, count in multi_presence.items():
    print(f"  {ds}: {count} files")
print()

# ============================================================================
# SECTION 3: Detailed Issue Report
# ============================================================================

print("\nSection 3: Detailed Issue Report")
print("=" * 50 + "\n")

problematic = [r for r in audit_results if r['status'] in ('ISSUES', 'ERROR', 'EMPTY')]

print(f"Found {len(problematic)} files with issues\n")

for i, item in enumerate(problematic, 1):
    print(f"[{i}] {item['file']}")
    print(f"    Status: {item['status']}")
    print(f"    Size: {item['n_rows']} rows x {item['n_cols']} cols")
    print("    Issues:")
    for issue in item['issues']:
        print(f"      - {issue}")
    print()

# ============================================================================
# SECTION 4: Generate Fix Recipes
# ============================================================================

print("\nSection 4: Generating Fix Recipes")
print("=" * 40 + "\n")

fix_recipes = []

for item in problematic:
    recipe = {
        'file': item['file'],
        'path': item['path'],
        'fixes': []
    }

    for issue in item['issues']:
        if 'placeholder treatment' in issue:
            recipe['fixes'].append("Load from source package and extract treatment labels")
        elif 'numeric/placeholder study IDs' in issue:
            recipe['fixes'].append("Load from source package and extract study names")
        elif 'duplicate' in issue:
            recipe['fixes'].append("Remove duplicate (study, treatment) rows, keep first")
        elif 'empty values' in issue or 'empty treatment' in issue or 'empty study' in issue:
            recipe['fixes'].append("Impute or remove rows with empty values")
        elif 'File is empty' in issue:
            recipe['fixes'].append("Re-extract from source package")

    if recipe['fixes']:
        fix_recipes.append(recipe)

print(f"Generated {len(fix_recipes)} fix recipes\n")

# ============================================================================
# SECTION 5: Save Results
# ============================================================================

print("\nSection 5: Saving Results")
print("=" * 40 + "\n")

# Create output directory
OUTPUT_DIR.mkdir(parents=True, exist_ok=True)

# Save audit summary
with open(OUTPUT_DIR / "data_audit_summary.csv", 'w', newline='', encoding='utf-8') as f:
    writer = csv.writer(f)
    writer.writerow(['file', 'status', 'n_rows', 'n_cols', 'n_issues', 'issues'])
    for r in audit_results:
        writer.writerow([
            r['file'],
            r['status'],
            r['n_rows'],
            r['n_cols'],
            len(r['issues']),
            '; '.join(r['issues'])
        ])
print("Saved: data_audit_summary.csv")

# Save problematic files detail
if problematic:
    with open(OUTPUT_DIR / "problematic_files_detail.csv", 'w', newline='', encoding='utf-8') as f:
        writer = csv.writer(f)
        writer.writerow(['file', 'status', 'issues'])
        for r in problematic:
            writer.writerow([
                r['file'],
                r['status'],
                '; '.join(r['issues'])
            ])
    print("Saved: problematic_files_detail.csv")

# Save fix recipes
if fix_recipes:
    with open(OUTPUT_DIR / "fix_recipes.csv", 'w', newline='', encoding='utf-8') as f:
        writer = csv.writer(f)
        writer.writerow(['file', 'fixes'])
        for r in fix_recipes:
            writer.writerow([
                r['file'],
                '; '.join(r['fixes'])
            ])
    print("Saved: fix_recipes.csv")

# Save cross-package dataset mapping
with open(OUTPUT_DIR / "cross_package_datasets.csv", 'w', newline='', encoding='utf-8') as f:
    writer = csv.writer(f)
    writer.writerow(['dataset', 'n_files'])
    for ds, count in sorted(dataset_presence.items(), key=lambda x: -x[1]):
        writer.writerow([ds, count])
print("Saved: cross_package_datasets.csv")

print("\n" + "="*73)
print("  AUDIT COMPLETE")
print("="*73)
print(f"\nResults saved to: {OUTPUT_DIR}")
print("\nSUMMARY:")
print(f"  Total files audited: {len(audit_results)}")
print(f"  Files with issues: {len(problematic)}")
print(f"  Cross-package datasets: {sum(1 for v in dataset_presence.values() if v > 2)}")
print("\nNext steps:")
print("  1. Review problematic_files_detail.csv")
print("  2. Apply fixes from fix_recipes.csv")
print("  3. Run cross-package comparison on clean datasets")
print()
